runtime/lua/vim/lsp/on_type_formatting.lua

```diff
@@ -0,0 +1,272 @@
+    if not next(trigger) then
+      buf_handle[i] = nil
+    end
+  end
+
+  -- Remove the buf handle and its autocmds if we removed its last client.
+  if not next(buf_handle) then
+    buf_handles[bufnr] = nil
+    api.nvim_clear_autocmds({ group = get_buf_augroup(bufnr) })
+
+    -- Remove the on_key callback if we removed the last buf handle.
+    if not next(buf_handles) then
+      vim.on_key(nil, type_formatting_ns)
+    end
+  end
+end
+
+--- @param client vim.lsp.Client
+--- @param bufnr integer
+local function attach_otf(client, bufnr)
+  local client_id = client.id
+  ---@type lsp.DocumentOnTypeFormattingOptions?
+  local otf_capabilities =
+    vim.tbl_get(client.server_capabilities, 'documentOnTypeFormattingProvider')
+  if not otf_capabilities then
+    vim.notify(
+      string.format(
+        'Client with id %d does not support textDocument/onTypeFormatting requests',
+        client_id
+      ),
+      vim.log.levels.WARN
+    )
+    return
+  end
+
+  -- Set on_key callback, clearing first in case it was already registered.
+  vim.on_key(nil, type_formatting_ns)
+  vim.on_key(on_key_cb, type_formatting_ns)
+
+  -- Populate the buf handle data. We cannot use defaulttable here because then an empty table will
+  -- be created for each unique keystroke
+  local buf_handle = buf_handles[bufnr] or {}
+  buf_handles[bufnr] = buf_handle
+
+  local trigger = buf_handle[otf_capabilities.firstTriggerCharacter] or {}
+  buf_handle[otf_capabilities.firstTriggerCharacter] = trigger
+  trigger[client_id] = client
+
+  for _, char in ipairs(otf_capabilities.moreTriggerCharacter or {}) do
+    trigger = buf_handle[char] or {}
+    buf_handle[char] = trigger
+    trigger[client_id] = client
+  end
+
+  local group = api.nvim_create_augroup(get_buf_augroup(bufnr), { clear = true })
+  api.nvim_create_autocmd('LspDetach', {
+    buffer = bufnr,
+    desc = 'Detach on-type formatting module when the client detaches',
+    group = group,
+    callback = function(args)
+      detach_otf(args.data.client_id, bufnr)
+    end,
+  })
+end
+
+---@param enable boolean
+---@param client vim.lsp.Client
+local function toggle_otf_for_client(enable, client)
+  local handler = enable and attach_otf or detach_otf
+  local client_id = client.id
+
+  -- Toggle for buffers already attached.
+  for bufnr, _ in pairs(client.attached_buffers) do
+    handler(client, bufnr)
+  end
+
+  -- If disabling, only clear the attachment autocmd. If enabling, create it.
+  local group = api.nvim_create_augroup(get_client_augroup(client_id), { clear = true })
+  if enable then
+    api.nvim_create_autocmd('LspAttach', {
+      group = group,
+      desc = 'Enable on-type formatting for all buffers this client attaches to',
+      callback = function(ev)
+        if ev.data.client_id ~= client_id then
+          return
+        end
+
+        attach_otf(client, ev.buf)
+      end,
+    })
+  end
+end
+
+---@param enable boolean
+local function toggle_otf_globally(enable)
```

Comment:
As this is a local function, the context already implies this is `otf`.


```suggestion
local function toggle_globally(enable)
```

But maybe just have a single `toggle()` function to handle both cases?